Fix controller client close issue while creating scope in multiController system test

**Problem description**
The changes in PR #2542 introduced unintentional closing of controller client, before the `createScope` future returns. This results in 
```
Caused by: java.util.concurrent.CompletionException: java.util.concurrent.RejectedExecutionException: 
```

**Problem location**
https://github.com/pravega/pravega/blob/master/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java#L179

**Suggestions for an improvement**
Close the controller client only after ensuring that scope is created successfully.

--- a/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
+++ b/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
@@ -27,7 +27,6 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 import lombok.Cleanup;

--- a/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
+++ b/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
@@ -27,8 +27,10 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
+import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
 import mesosphere.marathon.client.MarathonException;
 import org.apache.commons.lang3.RandomStringUtils;
@@ -44,7 +46,6 @@ import static io.pravega.test.system.framework.Utils.DOCKER_BASED;
 public class MultiControllerTest {
     private static final ScheduledExecutorService EXECUTOR_SERVICE = Executors.newSingleThreadScheduledExecutor();
     private Service controllerService = null;
-    private ControllerImpl controllerClient = null;
     private AtomicReference<URI> controllerURIDirect = new AtomicReference<>();
     private AtomicReference<URI> controllerURIDiscover = new AtomicReference<>();
 
@@ -140,15 +141,14 @@ public class MultiControllerTest {
 
         log.info("Test tcp:// with no controller instances running");
         AssertExtensions.assertThrows("Should throw RetriesExhaustedException",
-                createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()),
+                () -> createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()),
                 throwable -> throwable instanceof RetriesExhaustedException);
-        controllerClient.close();
+
         if (!DOCKER_BASED) {
             log.info("Test pravega:// with no controller instances running");
             AssertExtensions.assertThrows("Should throw RetriesExhaustedException",
-                    createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()),
+                    () -> createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()),
                     throwable -> throwable instanceof RetriesExhaustedException);
-            controllerClient.close();
         }
 
         log.info("multiControllerTest execution completed");
@@ -156,33 +156,41 @@ public class MultiControllerTest {
 
     private void withControllerURIDirect() throws ExecutionException, InterruptedException {
         Assert.assertTrue(createScopeWithSimpleRetry(
-                "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()).get());
-        controllerClient.close();
+                "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()));
     }
 
     private void withControllerURIDiscover() throws ExecutionException, InterruptedException {
         if (!DOCKER_BASED) {
             Assert.assertTrue(createScopeWithSimpleRetry(
-                    "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()).get());
+                    "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()));
         }
-        controllerClient.close();
 
     }
 
-    CompletableFuture<Boolean> createScopeWithSimpleRetry(String scopeName, URI controllerURI) {
+    private boolean createScopeWithSimpleRetry(String scopeName, URI controllerURI) throws ExecutionException, InterruptedException {
         // Need to retry since there is a delay for the mesos DNS name to resolve correctly.
-        return Retry.withExpBackoff(500, 2, 10, 5000)
-                .retryingOn(Exception.class)
-                .throwingOn(IllegalArgumentException.class)
-                .runAsync(() -> createScope(scopeName, controllerURI), EXECUTOR_SERVICE);
-    }
-
-    private CompletableFuture<Boolean> createScope(String scopeName, URI controllerURI) {
-        controllerClient = new ControllerImpl(ControllerImplConfig.builder()
+        @Cleanup
+        final ControllerImpl controllerClient = new ControllerImpl(ControllerImplConfig.builder()
                 .clientConfig(ClientConfig.builder()
                         .controllerURI(controllerURI)
                         .build())
                 .build(), EXECUTOR_SERVICE);
-        return controllerClient.createScope(scopeName);
+
+        CompletableFuture<Boolean> retryResult = Retry.withExpBackoff(500, 2, 10, 5000)
+                .retryingOn(Exception.class)
+                .throwingOn(IllegalArgumentException.class)
+                .runAsync(() -> controllerClient.createScope(scopeName), EXECUTOR_SERVICE);
+
+        return retryResult.get();
+    }
+
+    private boolean createScope(String scopeName, URI controllerURI) throws ExecutionException, InterruptedException {
+        @Cleanup
+        final ControllerImpl controllerClient = new ControllerImpl(ControllerImplConfig.builder()
+                .clientConfig(ClientConfig.builder()
+                        .controllerURI(controllerURI)
+                        .build())
+                .build(), EXECUTOR_SERVICE);
+        return controllerClient.createScope(scopeName).get();
     }
 }

--- a/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
+++ b/test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java
@@ -29,8 +29,6 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
-
-import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
 import mesosphere.marathon.client.MarathonException;
 import org.apache.commons.lang3.RandomStringUtils;
@@ -39,7 +37,6 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-
 import static io.pravega.test.system.framework.Utils.DOCKER_BASED;
 
 @Slf4j
@@ -47,6 +44,7 @@ import static io.pravega.test.system.framework.Utils.DOCKER_BASED;
 public class MultiControllerTest {
     private static final ScheduledExecutorService EXECUTOR_SERVICE = Executors.newSingleThreadScheduledExecutor();
     private Service controllerService = null;
+    private ControllerImpl controllerClient = null;
     private AtomicReference<URI> controllerURIDirect = new AtomicReference<>();
     private AtomicReference<URI> controllerURIDiscover = new AtomicReference<>();
 
@@ -144,11 +142,13 @@ public class MultiControllerTest {
         AssertExtensions.assertThrows("Should throw RetriesExhaustedException",
                 createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()),
                 throwable -> throwable instanceof RetriesExhaustedException);
+        controllerClient.close();
         if (!DOCKER_BASED) {
             log.info("Test pravega:// with no controller instances running");
             AssertExtensions.assertThrows("Should throw RetriesExhaustedException",
                     createScope("scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()),
                     throwable -> throwable instanceof RetriesExhaustedException);
+            controllerClient.close();
         }
 
         log.info("multiControllerTest execution completed");
@@ -157,6 +157,7 @@ public class MultiControllerTest {
     private void withControllerURIDirect() throws ExecutionException, InterruptedException {
         Assert.assertTrue(createScopeWithSimpleRetry(
                 "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDirect.get()).get());
+        controllerClient.close();
     }
 
     private void withControllerURIDiscover() throws ExecutionException, InterruptedException {
@@ -164,6 +165,7 @@ public class MultiControllerTest {
             Assert.assertTrue(createScopeWithSimpleRetry(
                     "scope" + RandomStringUtils.randomAlphanumeric(10), controllerURIDiscover.get()).get());
         }
+        controllerClient.close();
 
     }
 
@@ -176,8 +178,7 @@ public class MultiControllerTest {
     }
 
     private CompletableFuture<Boolean> createScope(String scopeName, URI controllerURI) {
-        @Cleanup
-        final ControllerImpl controllerClient = new ControllerImpl(ControllerImplConfig.builder()
+        controllerClient = new ControllerImpl(ControllerImplConfig.builder()
                 .clientConfig(ClientConfig.builder()
                         .controllerURI(controllerURI)
                         .build())

